File: expo-app/src/lib/parsers/redfinUrlParser.ts
/**
 * Parses Redfin desktop, mobile, and common URL shapes. Extracts listing ID and address.
 */

import type { PartialAddress, ParsedListingURL, ParseURLResult } from './types';
import { slugToAddressLine } from './slugToAddressLine';
import { getHostname } from './listingUrlNormalize';
import { isRedfinHost } from './listingProviders';

function getQueryParam(u: URL, key: string): string | null {
  return u.searchParams.get(key);
}

function partialFromSlug(slug: string): PartialAddress | null {
  const line = slugToAddressLine(slug);
  if (!line) return null;
  const parts = line.split(',').map((p) => p.trim());
  if (parts.length >= 3) {
    const tail = parts[parts.length - 1]?.split(/\s+/) ?? [];
    const zip = tail.find((x) => /^\d{5}/.test(x)) ?? null;
    const state = tail.find((x) => /^[A-Z]{2}$/.test(x)) ?? null;
    return {
      streetAddress: parts[0] ?? null,
      city: parts[1] ?? null,
      state,
      postalCode: zip,
    };
  }
  return { streetAddress: line, city: null, state: null, postalCode: null };
}

/**
 * Redfin paths often: /ST/City/street-zip/home/id or /ST/City/address-slug/home/id
 */
function extractListingIdFromComponents(components: string[]): string | null {
  const homeIdx = components.findIndex((c) => c.toLowerCase() === 'home');
  if (homeIdx >= 0 && components[homeIdx + 1] && /^\d+$/.test(components[homeIdx + 1])) {
    return components[homeIdx + 1];
  }
  const last = components[components.length - 1];
  if (last && /^\d+$/.test(last) && components.length >= 2) {
    const prev = components[components.length - 2]?.toLowerCase();
    if (prev === 'home' || prev === 'unit') return last;
  }
  if (last && /^\d{6,12}$/.test(last)) return last;
  return null;
}

function findAddressSlug(components: string[]): string | null {
  const homeIdx = components.findIndex((c) => c.toLowerCase() === 'home');
  if (homeIdx > 0) {
    const candidate = components[homeIdx - 1];
    if (candidate?.includes('-') && candidate.length > 5) return candidate;
  }
  const slug = components.find(
    (c) => c.includes('-') && c.length > 5 && !/^\d+$/.test(c) && c.toLowerCase() !== 'home'
  );
  return slug ?? null;
}

export function parseRedfinUrl(urlString: string): ParseURLResult {
  const host = getHostname(urlString);
  if (!host || !isRedfinHost(host)) {
    return { ok: false, error: { kind: 'unsupportedDomain' } };
  }

  try {
    const u = new URL(urlString);
    let listingID: string | null =
      getQueryParam(u, 'listingId') ??
      getQueryParam(u, 'listing_id') ??
      getQueryParam(u, 'listingID');

    const components = u.pathname.replace(/\/$/, '').split('/').filter(Boolean);

    if (!listingID) {
      listingID = extractListingIdFromComponents(components);
    }

    if (!listingID || listingID.length === 0) {
      return { ok: false, error: { kind: 'missingListingID' } };
    }

    const slug = findAddressSlug(components);
    const address = slug ? partialFromSlug(slug) : null;

    return {
      ok: true,
      value: {
        source: 'redfin',
        listingID,
        address,
        originalURL: urlString,
      },
    };
  } catch {
    return { ok: false, error: { kind: 'missingListingID' } };
  }
}

export function redfinAddressLineForImport(parsed: ParsedListingURL): string | null {
  try {
    const u = new URL(parsed.originalURL);
    const components = u.pathname.split('/').filter(Boolean);
    const slug = findAddressSlug(components);
    if (slug) {
      const line = slugToAddressLine(slug);
      if (line.trim()) return line.trim();
    }
  } catch {
    /* ignore */
  }
  const a = parsed.address;
  if (a) {
    const parts = [a.streetAddress, a.city, a.state, a.postalCode].filter(Boolean);
    if (parts.length > 0) return parts.join(', ');
  }
  return null;
}

File: PropFolio/Services/PropertyData/Parsers/RedfinURLParser.swift
//
//  RedfinURLParser.swift
//  PropFolio
//
//  Parses Redfin desktop and mobile URLs. Extracts listing ID (and optional address).
//

import Foundation

struct RedfinURLParser: ListingURLParser {
    private let desktopHosts = ["www.redfin.com", "redfin.com"]
    private let mobileHosts = ["m.redfin.com", "redfin.com"]

    func canParse(url: URL) -> Bool {
        guard let host = url.host?.lowercased() else { return false }
        return desktopHosts.contains(host) || host.hasPrefix("m.") && host.hasSuffix("redfin.com")
    }

    func parse(url: URL) -> Result<ParsedListingURL, URLParseError> {
        guard canParse(url: url) else { return .failure(.unsupportedDomain) }

        // Desktop: /TX/Austin/123-Main-St-78701/unit/1234567890 or /home/1234567890
        // Mobile:  similar path or /listing/1234567890
        // ID is often in path as numeric segment or in query "listingId"
        var listingID: String?

        if let id = url.queryItems?["listingId"] ?? url.queryItems?["listing_id"] {
            listingID = id
        }

        let path = url.path
        let components = path.split(separator: "/").map(String.init)

        if listingID == nil {
            // Last path component is often the listing ID (numeric)
            if let last = components.last, last.allSatisfy({ $0.isNumber }) {
                listingID = last
            }
            // Or second-to-last for .../unit/1234567890
            if listingID == nil, components.count >= 2, components[components.count - 2].lowercased() == "unit",
               components.last?.allSatisfy({ $0.isNumber }) == true {
                listingID = components.last
            }
        }

        guard let id = listingID, !id.isEmpty else { return .failure(.missingListingID) }

        let partial = extractAddressFromPath(components)
        return .success(ParsedListingURL(
            source: .redfin,
            listingID: id,
            address: partial,
            originalURL: url
        ))
    }

    private func extractAddressFromPath(_ pathComponents: [String]) -> PartialAddress? {
        // /STATE/City/Street-Zip/... or /STATE/City/...
        guard pathComponents.count >= 3 else { return nil }
        let state = pathComponents[0].uppercased()
        let city = pathComponents[1].replacingOccurrences(of: "-", with: " ")
        var street: String?
        var postalCode: String?
        if pathComponents.count >= 4 {
            let slug = pathComponents[2]
            let parts = slug.split(separator: "-").map(String.init)
            if let last = parts.last, last.count == 5, last.allSatisfy({ $0.isNumber }) {
                postalCode = last
                street = parts.dropLast().joined(separator: " ").replacingOccurrences(of: "-", with: " ")
            } else {
                street = slug.replacingOccurrences(of: "-", with: " ")
            }
        }
        return PartialAddress(streetAddress: street, city: city, state: state, postalCode: postalCode)
    }
}

private extension URL {
    var queryItems: [String: String]? {
        guard let comp = URLComponents(url: self, resolvingAgainstBaseURL: false), let items = comp.queryItems else { return nil }
        return Dictionary(uniqueKeysWithValues: items.map { ($0.name, $0.value ?? "") })
    }
}
